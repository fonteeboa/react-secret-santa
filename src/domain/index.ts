export interface Participant {
    name: string;
}

export interface PlayerFormProps {
    onAddParticipant: (participant: Participant) => void;
}

export interface PlayerListProps {
    participants: Participant[];
    onRemoveParticipant: (index: number) => void;
}

export interface ResultListProps {
    assignments: Map<Participant, Participant>;
}
