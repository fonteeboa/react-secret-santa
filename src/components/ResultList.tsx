import React from 'react';
import { List } from 'antd';
import VirtualList from 'rc-virtual-list';
import { Participant, ResultListProps } from '../domain';
import { useTranslation } from 'react-i18next';

const ResultList: React.FC<ResultListProps> = ({ assignments }) => {
    const { t } = useTranslation();
    const ContainerHeight = 627;
    const data = Array.from(assignments.entries());

    return (
        <List bordered className='resultList'>
            <VirtualList
                data={data}
                height={ContainerHeight}
                itemHeight={47}
                itemKey={(item) => `${item[0].name}-${item[1].name}`}
            >
                {(item: [Participant, Participant]) => (
                    <List.Item key={`${item[0].name}-${item[1].name}`}>
                        {t('give_gift_to', { giver: item[0].name, receiver: item[1].name })}
                    </List.Item>
                )}
            </VirtualList>
        </List>
    );
};

export default ResultList;
