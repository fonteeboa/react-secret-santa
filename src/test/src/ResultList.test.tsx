import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import ResultList from '../../components/ResultList';
import { Participant } from '../../domain';
import { useTranslation } from 'react-i18next';
import '../mocks/matchMediaMock';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string, params: any) => {
      if (key === 'give_gift_to') {
        return `${params.giver} will give a gift to ${params.receiver}`;
      }
      return key;
    },
  }),
}));

describe('ResultList Component', () => {
  const participant1: Participant = { name: 'John' };
  const participant2: Participant = { name: 'Jane' };
  const participant3: Participant = { name: 'Doe' };
  
  const assignments = new Map<Participant, Participant>([
    [participant1, participant2],
    [participant2, participant3],
    [participant3, participant1],
  ]);

  test('renders the result list correctly', () => {
    render(<ResultList assignments={assignments} />);

    expect(screen.getByText('John will give a gift to Jane')).toBeInTheDocument();
    expect(screen.getByText('Jane will give a gift to Doe')).toBeInTheDocument();
    expect(screen.getByText('Doe will give a gift to John')).toBeInTheDocument();
  });

  test('renders an empty result list correctly', () => {
    const emptyAssignments = new Map<Participant, Participant>();
    render(<ResultList assignments={emptyAssignments} />);

    expect(screen.queryByText('will give a gift to')).not.toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const { asFragment } = render(<ResultList assignments={assignments} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
