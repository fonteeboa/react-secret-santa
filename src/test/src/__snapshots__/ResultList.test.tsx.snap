// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultList Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-list ant-list-split ant-list-bordered resultList css-dev-only-do-not-override-1uq9j6g"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-1uq9j6g"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="rc-virtual-list"
          style="position: relative;"
        >
          <div
            class="rc-virtual-list-holder"
            style="height: 627px; overflow-y: hidden;"
          >
            <div>
              <div
                class="rc-virtual-list-holder-inner"
                style="display: flex; flex-direction: column;"
              >
                <li
                  class="ant-list-item"
                >
                  John will give a gift to Jane
                </li>
                <li
                  class="ant-list-item"
                >
                  Jane will give a gift to Doe
                </li>
                <li
                  class="ant-list-item"
                >
                  Doe will give a gift to John
                </li>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
